devito/operations/interpolators.py

```diff
@@ -108,108 +107,111 @@ def interpolate(self, *args, **kwargs):
+
     @cached_property
+    def _psym(self):
+        return self.sfunction._point_symbols
+
+    @cached_property
+    def _gdim(self):
+        return self.grid.dimensions
+
+    def implicit_dims(self, implicit_dims):
+        return as_tuple(implicit_dims) + self.sfunction.dimensions
+
+    @cached_property
+    def r(self):
+        return self.sfunction.r
+
+    @property
+    def _interp_points(self):
+        return range(-self.r+1, self.r+1)
+
+    @property
+    def _nd_points(self):
+        return product(self._interp_points, repeat=self.grid.dim)
+
+    @property
     def _interpolation_coeffs(self):
-        """
-        Symbolic expression for the coefficients for sparse point interpolation
-        according to:
+        coeffs = {}
+        for p in self._nd_points:
+            coeffs[p] = prod([self._weights[d][i] for (d, i) in zip(self._gdim, p)])
+        return list(coeffs.values())
 
-            https://en.wikipedia.org/wiki/Bilinear_interpolation.
+    def _coeff_temps(self, implicit_dims):
+        return []
 
-        Returns
-        -------
-        Matrix of coefficient expressions.
-        """
-        # Grid indices corresponding to the corners of the cell ie x1, y1, z1
-        indices1 = tuple(sympy.symbols('%s1' % d) for d in self.grid.dimensions)
-        indices2 = tuple(sympy.symbols('%s2' % d) for d in self.grid.dimensions)
-        # 1, x1, y1, z1, x1*y1, ...
-        indices = list(powerset(indices1))
-        indices[0] = (1,)
-        point_sym = list(powerset(self.sfunction._point_symbols))
-        point_sym[0] = (1,)
-        # 1, px. py, pz, px*py, ...
-        A = []
-        ref_A = [np.prod(ind) for ind in indices]
-        # Create the matrix with the same increment order as the point increment
-        for i in self.sfunction._point_increments:
-            # substitute x1 by x2 if increment in that dimension
-            subs = dict((indices1[d], indices2[d] if i[d] == 1 else indices1[d])
-                        for d in range(len(i)))
-            A += [[1] + [a.subs(subs) for a in ref_A[1:]]]
-
-        A = sympy.Matrix(A)
-        # Coordinate values of the sparse point
-        p = sympy.Matrix([[np.prod(ind)] for ind in point_sym])
-
-        # reference cell x1:0, x2:h_x
-        left = dict((a, 0) for a in indices1)
-        right = dict((b, dim.spacing) for b, dim in zip(indices2, self.grid.dimensions))
-        reference_cell = {**left, **right}
-        # Substitute in interpolation matrix
-        A = A.subs(reference_cell)
-        return A.inv().T * p
+    def _positions(self, implicit_dims):
+        return [Eq(v, k, implicit_dims=implicit_dims)
+                for k, v in self.sfunction._position_map.items()]
+
+    def subs_coords(self, _expr, *idx_subs):
+        return [_expr.xreplace(v_sub) * b.xreplace(v_sub)
+                for b, v_sub in zip(self._interpolation_coeffs, idx_subs)]
+
+    def subs_coords_eq(self, field, _expr, *idx_subs, implicit_dims=None):
```

Comment:
**if** they're called in just **one** place, probably no need for a separate method. Regardless, it's a private method, so it should with `_`